compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -605,15 +605,23 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
     def subst(from: List[Symbol], to: List[Symbol])(implicit ctx: Context): ThisTree =
       new TreeTypeMap(substFrom = from, substTo = to).apply(tree)
 
-    /** Ensure that top-level definitions of the current tree have the given owner */
+    /** Ensure that top-level definitions of the current tree have the given owner
+     *
+     *  @note Dependeing on the shape of the tree, this operation can be expensive.
+     *        Always prefer using `changeOwner` directly.
+     */
     def ensureOwner(owner: Symbol)(implicit ctx: Context): tpd.Tree = {
```

Comment:
In fact, there's a more efficient way to do it my using TreeTypeMap directly. This can replace multiple owners in one traversal. Then you won't need the @note anymore.